src/EditorFeatures/Core/Shared/Preview/PreviewSolutionCrawlerRegistrationService.cs

```diff
@@ -68,6 +74,34 @@ public void Unregister(Workspace workspace, bool blockingShutdown = false)
             {
                 source.Cancel();
             }
+
+            ClearDiagnostics(workspace);
+
+            AssertCleanup();
+        }
+
+        private void ClearDiagnostics(Workspace workspace)
+        {
+            var analyzer = _provider.CreateIncrementalAnalyzer(workspace);
+            var args = _service.GetDiagnosticsUpdatedEventArgs(workspace, projectId: null, documentId: null, cancellationToken: CancellationToken.None);
+
+            foreach (var documentId in args.Select(a => a.DocumentId).Distinct())
+            {
+                analyzer.RemoveDocument(documentId);
```

Comment:
It seems like a much cleaner model would be to have Register(Workspace) just keep a dictionary around that maps from the Workspace to associated analyzer that is created for it.  Then Unregister can just collect that analyzer and ask it to clear up any of it's state.

Note: i think it should do this by just calling into the analyzer itself and saying "Dump all data associated with this workspace".  It shouldn't do things like call "RemoveDocument".  That type of approach means that this code here has to understand far too deeply how an analyzer caches data and how that data can be fully cleaned up.  The analyzer should know how to do that, that way all the caching and clearing logic is in one place.